### PR TITLE
SONARJAVA-1256 Issues should be saved on the go or by file instead of using a project index

### PR DESCRIPTION
**Pull request by mpaladin**
https://jira.sonarsource.com/browse/SONARJAVA-1256

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/ast/JavaAstScanner.java [hunk 1]
```diff
@@ -129,8 +132,12 @@ public SourceCodeSearchEngine getIndex() {
 
   /**
    * Helper method for testing checks without having to deploy them on a Sonar instance.
+   * Can be dropped when support for CheckMessageVerifier will be dropped.
+   *
+   * @deprecated As of release 3.6, should use {@link org.sonar.java.checks.verifier.JavaCheckVerifier#verify(String filename, JavaFileScanner check)} for rules unit tests.
    */
   @VisibleForTesting
+  @Deprecated
   public static SourceFile scanSingleFile(File file, VisitorsBridge visitorsBridge) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/ast/JavaAstScanner.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: java-squid/src/main/java/org/sonar/java/JavaSquid.java [hunk 2]
```diff
@@ -104,16 +98,17 @@ public JavaSquid(JavaConfiguration conf,
...
+    bytecodeScanner.accept(dependenciesVisitor);
     for (CodeVisitor visitor : visitors) {
       bytecodeScanner.accept(visitor);
     }
 
   }
 
-  private static VisitorsBridge createVisitorBridge(Iterable<CodeVisitor> codeVisitors, List<File> classpath, JavaConfiguration conf, @Nullable SonarComponents sonarComponents) {
-    VisitorsBridge visitorsBridge = new VisitorsBridge(codeVisitors, classpath, sonarComponents);
+  private static InternalVisitorsBridge createVisitorBridge(Iterable<CodeVisitor> codeVisitors, List<File> classpath, JavaConfiguration conf, @Nullable SonarComponents sonarComponents) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/JavaSquid.java]: MINOR Split this 186 characters long line (which is greater than 180 authorized). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00103)

File: java-squid/src/main/java/org/sonar/java/model/InternalVisitorsBridge.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    }
+  }
+
+  protected JavaFileScannerContext createScannerContext(CompilationUnitTree tree, SemanticModel semanticModel, boolean analyseAccessors, SonarComponents sonarComponents) {
+    return new DefaultJavaFileScannerContext(tree, (SourceFile) getContext().peekSourceCode(), getContext().getFile(), semanticModel, analyseAccessors, sonarComponents);
+  }
+
+  private boolean isNotJavaLangOrSerializable(String packageName) {
+    String name = getContext().getFile().getName();
+    return !("java/lang".equals(packageName)
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/InternalVisitorsBridge.java]: MAJOR Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)

File: java-squid/src/main/java/org/sonar/java/model/VisitorsBridge.java [hunk 4]
```diff
@@ -86,141 +64,44 @@ public VisitorsBridge(JavaFileScanner visitor, List<File> projectClasspath) {
...
 
-    @Override
-    public void addIssue(int line, JavaCheck javaCheck, String message, @Nullable Double cost) {
-      Preconditions.checkNotNull(javaCheck);
-      Preconditions.checkNotNull(message);
+    /**
+     * @deprecated should be removed once CheckMessageVerifier support is dropped
+     */
+    @Deprecated
+    private void addIssueForCheckMessageVerifier(int line, JavaCheck javaCheck, String message, @Nullable Double cost) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/VisitorsBridge.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: java-squid/src/main/java/org/sonar/java/model/VisitorsBridge.java [hunk 5]
```diff
@@ -86,141 +64,44 @@ public VisitorsBridge(JavaFileScanner visitor, List<File> projectClasspath) {
...
+    public Set<AnalyzerMessage> getIssues() {
+      return issues;
     }
 
     @Override
-    public void addIssue(int line, JavaCheck javaCheck, String message) {
-      addIssue(line, javaCheck, message, null);
+    public void addIssue(int line, JavaCheck javaCheck, String message, @Nullable Double cost) {
+      issues.add(new AnalyzerMessage(javaCheck, getFile(), line, message, cost != null ? cost.intValue() : 0));
+      addIssueForCheckMessageVerifier(line, javaCheck, message, cost);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/VisitorsBridge.java]: MINOR Method 'TestJavaFileScannerContext.addIssueForCheckMessageVerifier(...)' is deprecated. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: java-squid/src/main/java/org/sonar/java/model/VisitorsBridge.java [hunk 6]
```diff
@@ -238,120 +119,16 @@ public void addIssue(int line, JavaCheck javaCheck, String message, @Nullable Do
...
-      } else {
-        addIssue(syntaxNode, javaCheck, message, cost != null ? (double) cost : null);
+      File file = getFile();
+      AnalyzerMessage analyzerMessage = new AnalyzerMessage(javaCheck, file, textSpanFor(syntaxNode), message, cost != null ? cost : 0);
+      for (Location location : secondary) {
+        AnalyzerMessage secondaryLocation = new AnalyzerMessage(javaCheck, file, textSpanFor(location.syntaxNode), location.msg, 0);
+        analyzerMessage.secondaryLocations.add(secondaryLocation);
       }
+      issues.add(analyzerMessage);
+      addIssueForCheckMessageVerifier(analyzerMessage.getLine(), javaCheck, message, (double) (cost != null ? cost : 0));
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/model/VisitorsBridge.java]: MINOR Method 'TestJavaFileScannerContext.addIssueForCheckMessageVerifier(...)' is deprecated. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 18 issues:
* CRITICAL 4 critical
* MAJOR 9 major
* MINOR 3 minor
* INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* CRITICAL [Remove this use of dynamic class loading.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L67) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2658)
* MAJOR [Rename this field "issueBuilder_at" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L54) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* MAJOR [Rename this field "newIssueLocation_message" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L52) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* CRITICAL [Remove this use of dynamic class loading.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L69) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2658)
* MAJOR [Rename this field "issueBuilder_addFlow" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L56) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* MAJOR [Rename this field "inputFile_selectLine" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L48) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* CRITICAL [Either log or rethrow this exception.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L61) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)
* MAJOR [Rename this field "newIssueLocation_on" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L50) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* CRITICAL [Remove this use of dynamic class loading.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L68) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2658)
* MAJOR [Rename this field "issueBuilder_newLocation" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L55) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* MAJOR [Rename this field "newIssueLocation_at" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L51) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)
* MAJOR [Rename this field "inputFile_newRange" to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/SonarSource/sonar-java/blob/9e84af2c0045603758b0096b5cef61b151350356/java-squid/src/main/java/org/sonar/java/CompIssue.java#L47) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00116)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MINOR 2 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MINOR 2 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues:
* MINOR 2 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by m-g-sonar** [comment]
I close this PR, as it is has been duplicated in https://github.com/SonarSource/sonar-java/pull/436 with splitted commits.

**Pull request closed without merging**